src/stream/src/executor/sink.rs

```diff
@@ -12,42 +12,76 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::HashMap;
+
 use futures::StreamExt;
 use futures_async_stream::try_stream;
 use risingwave_common::catalog::Schema;
-use risingwave_connector::sink::Sink;
+use risingwave_common::error::Result;
+use risingwave_connector::sink::{SinkConfig, SinkImpl};
+use risingwave_storage::StateStore;
 
 use super::error::StreamExecutorError;
 use super::{BoxedExecutor, Executor, Message};
 
-pub struct SinkExecutor<S: Sink> {
-    child: BoxedExecutor,
-    _external_sink: S,
+pub struct SinkExecutor<S: StateStore> {
+    input: BoxedExecutor,
+    _store: S,
+    properties: HashMap<String, String>,
     identity: String,
 }
 
-impl<S: Sink> SinkExecutor<S> {
-    pub fn _new(materialize_executor: BoxedExecutor, _external_sink: S) -> Self {
+async fn build_sink(config: SinkConfig) -> Result<Box<SinkImpl>> {
+    Ok(Box::new(SinkImpl::new(config).await?))
+}
+
+impl<S: StateStore> SinkExecutor<S> {
+    pub fn new(
+        materialize_executor: BoxedExecutor,
+        _store: S,
+        mut properties: HashMap<String, String>,
+        executor_id: u64,
+    ) -> Self {
+        // This field can be used to distinguish a specific actor in parallelism to prevent
+        // transaction execution errors
+        properties.insert("identifier".to_string(), format!("sink-{:?}", executor_id));
         Self {
-            child: materialize_executor,
-            _external_sink,
-            identity: "SinkExecutor".to_string(),
+            input: materialize_executor,
+            _store,
+            properties,
+            identity: format!("SinkExecutor_{:?}", executor_id),
         }
     }
 
     #[try_stream(ok = Message, error = StreamExecutorError)]
     async fn execute_inner(self) {
-        todo!()
+        let sink_config = SinkConfig::from_hashmap(self.properties.clone())
+            .map_err(StreamExecutorError::sink_error)?;
+        let _sink = build_sink(sink_config).await;
+        let input = self.input.execute();
+        #[for_await]
+        for msg in input {
+            match msg? {
+                Message::Chunk(chunk) => {
+                    let _visible_chunk = chunk.clone().compact()?;
+
+                    yield Message::Chunk(chunk);
+                }
+                Message::Barrier(barrier) => {
+                    yield Message::Barrier(barrier);
+                }
+            }
+        }
     }
 }
 
-impl<S: Sink + 'static + Send> Executor for SinkExecutor<S> {
+impl<S: StateStore> Executor for SinkExecutor<S> {
```

Comment:
Just for my understanding, this means that SinkExecuter is a stateful operator from now on?

Reply:
yes, we should store the inner state when barrier comes like other operators